### ISSUE TITLE
[C/C++] Using macro without semicolon breaks highlighting

### ISSUE THREAD

**Issue by pgrit** [sublimehq/Packages#1034]
Using a macro without a ; afterwards breaks the syntax highlighting of subsequent lines. See example below. 
![screenshot from 2017-06-15 18-56-48](https://user-images.githubusercontent.com/12607894/27192815-03bef236-51fd-11e7-8525-d7397f9484ba.png)

A similar behaviour can be observed for other (invalid) examples
![screenshot from 2017-06-15 19-06-01](https://user-images.githubusercontent.com/12607894/27193044-bebecbd8-51fd-11e7-8c0e-6980cf1a9497.png)

Maybe it is a good idea to treat new lines as semicolons for the sake of syntax highlighting?

**Comment by rwols** [comment]
> Maybe it is a good idea to treat new lines as semicolons for the sake of syntax highlighting?

I don't think that's a good idea, as that might pop off a carefully constructed context that was not intended to pop off.

In practice preprocessor tokens are UPPER_CASE, so perhaps one way to deal with this is to eat lines that *only* have UPPER_CASE identifiers (and nothing else) and continue parsing as normal.

In the meantime, you could keep the semicolon after the macro -- in theory redundant, but the generated code will be exactly the same as without the extra semicolon.

**Comment by wbond** [comment]
Yeah, I know there are bunch of bug reports that have come from users who split constructs across multiple lines with C and C++, so I don't think it will be a viable option to treat newlines as semicolons.

If you can paste some of the example code into fenced code blocks here, we can test with a recent version of the C++ syntax. (I am guessing you are using build 3126 since you still have the old default theme.)

**Comment by pgrit** [comment]
Yes, I am using build 3126

The sample code:

```C++
#define MY_NS_ENTER namespace my_ns {
#define MY_NS_EXIT }

MY_NS_ENTER

class X {
private:
    int a;

public:
    int b;
};

MY_NS_EXIT
```

One more thing that I noticed in my example just now: the "namespace" in the define itself is also not highlighted.

**Comment by rwols** [comment]
That's because the preprocessor is a textual thing and it's woefully unaware of the actual C/C++/Objective-C/Objective-C++ syntax. It's a deliberate decision because, for instance, line-continuation backslashes *are* highlighted (and are part of the syntax for the preprocessor):

```c
#define FOO bar \
            /*  ^ meta.preprocessor.macro punctuation.separator.continuation */ \
            baz
```
Go to Tools -> Developer -> Show Scope Name to see the scope names. 👍

**Comment by pgrit** [comment]
I found another, even more curious case. Here, the result is actually affected by the indentation and only happens inside a class or struct. The issue only occurs if there is whitespace at the beginning of the next line.

Note that in this example, eating UPPER_CASE would not work, and adding a semicolon would also not be possible.

![screenshot from 2017-06-17 16-19-44](https://user-images.githubusercontent.com/12607894/27253663-e162db12-5378-11e7-81a1-c4d88c5357ab.png)

```C++
#define DEPRECATED(msg) [[deprecated(msg)]]

struct Test {
    DEPRECATED("bla")
    bool foobarInBlue (bool run=true) {}
}

namespace Test {
    DEPRECATED("bla")
    bool foobarInBlue (bool run=true) {}
}

struct Test {
DEPRECATED("bla")
bool foobarInBlue (bool run=true) {}
}
```

**Comment by rwols** [comment]
I think it's feasible to also account for these kinds of cases. Give me some time to update my pull request.

**Comment by rwols** [comment]
All of this tricky stuff works correctly now.
